### DISCUSSION TITLE
Inconsistent colorspace

### DISCUSSION THREAD

**Discussion by elalish** [google/model-viewer#2200] [Q&A]
This looks like a browser bug, though I'm not sure who is right and who is wrong.

If we turn off OffscreenCanvas, then both Firefox and Chrome use the same codepaths (direct canvas rendering for a single instance, and drawImage transfers for multiple instances). In this case they are each self-consistent for single vs. multiple instances, but not with each other:
Chrome:
<img width="384" alt="ChromeNoOffscreen" src="https://user-images.githubusercontent.com/1649964/80404964-cfad9700-8876-11ea-8f2a-11c773d5aeac.png">
Firefox:
<img width="497" alt="FirefoxNoOffscreen" src="https://user-images.githubusercontent.com/1649964/80404981-d50ae180-8876-11ea-83a9-106bb39563b5.png">
Note the red buttons: for Chrome we get RGB = [198, 58, 58], while Firefox we get RGB = [215, 40, 50]. This is on a macbook Pro, so some kind of colorspace is getting incorrectly applied in one or the other.

Stranger yet, with OffscreenCanvas enabled (Chrome only), for single instance direct draw, we now get the same value as Firefox had, while for multiple instance using transferImageBitmap, we still get the non-OffscreenCanvas Chrome value. 

I'd love to have a test that would tell me definitively which one is correct, but I haven't thought of anything I would trust yet.

**Comment by kenrussell** [comment]
@jdashg do you think this could be related to Firefox's recent release of colorspace handling? There are several associated tests in the WebGL conformance suite, but I'm not sure whether this particular case is covered.

**Comment by kdashg** [comment]
It's hard to say, but possibly!
The name+info for the pref toggle is here:
https://searchfox.org/mozilla-central/rev/158bac3df3a1890da55bdb6ffdaf9a7ffc0bfb0a/modules/libpref/init/StaticPrefList.yaml#3686

It would be great to have a testcase that's 50%gray which would make it easier to tell (or at least guess) who's getting what wrong.

**Comment by kainino0x** [comment]
Yup, a test case (testing the various presentation paths) would be really helpful. BTW, it should include a css `#7f7f7f` - that would be our ground truth (within each browser, a 50%gray canvas should be the same color as 50% gray css). We probably wouldn't necessarily guarantee consistency across browsers as they may color manage differently.

**Comment by kdashg** [comment]
That's true, we can't guarantee a window-pixel result. However we do (should?) require consistency between elements in a page. 50% red should be the same coming from CSS or WebGL, though it may have different window tints across different browsers.
